obfuscate_and_normalize.go

```diff
@@ -5,40 +5,41 @@ import "strings"
 // ObfuscateAndNormalize takes an input SQL string and returns an normalized SQL string with metadata
 // This function is a convenience function that combines the Obfuscator and Normalizer in one pass
 func ObfuscateAndNormalize(input string, obfuscator *Obfuscator, normalizer *Normalizer, lexerOpts ...lexerOption) (normalizedSQL string, statementMetadata *StatementMetadata, err error) {
-	lexer := New(
-		input,
-		lexerOpts...,
-	)
-
-	var normalizedSQLBuilder strings.Builder
-
-	statementMetadata = &StatementMetadata{
-		Tables:     []string{},
-		Comments:   []string{},
-		Commands:   []string{},
-		Procedures: []string{},
-	}
+	lexer := New(input, lexerOpts...)
+	normalizedSQLBuilder := new(strings.Builder)
+	normalizedSQLBuilder.Grow(len(input))
+
+	// Always allocate metadata for backward compatibility
+	statementMetadata = statementMetadataPool.Get().(*StatementMetadata)
+	statementMetadata.reset()
+	defer statementMetadataPool.Put(statementMetadata)
 
-	var lastToken Token // The last token that is not whitespace or comment
 	var groupablePlaceholder groupablePlaceholder
 	var headState headState
+	var ctes map[string]bool
+
+	// Only allocate CTEs map if collecting tables
+	if normalizer.config.CollectTables {
+		ctes = make(map[string]bool, 2)
+	}
 
-	ctes := make(map[string]bool) // Holds the CTEs that are currently being processed
+	var lastValueToken *LastValueToken
 
 	for {
 		token := lexer.Scan()
-		token.Value = obfuscator.ObfuscateTokenValue(token, lastToken, lexerOpts...)
-		normalizer.collectMetadata(&token, &lastToken, statementMetadata, ctes)
-		normalizer.normalizeSQL(&token, &lastToken, &normalizedSQLBuilder, &groupablePlaceholder, &headState, lexerOpts...)
+		obfuscator.ObfuscateTokenValue(token, lastValueToken, lexerOpts...)
+		if normalizer.shouldCollectMetadata() {
+			normalizer.collectMetadata(token, lastValueToken, statementMetadata, ctes)
+		}
+		normalizer.normalizeSQL(token, lastValueToken, normalizedSQLBuilder, &groupablePlaceholder, &headState, lexerOpts...)
 		if token.Type == EOF {
 			break
 		}
+		if isValueToken(token) {
+			lastValueToken = token.getLastValueToken()
+		}
 	}
 
 	normalizedSQL = normalizedSQLBuilder.String()
-
-	// Dedupe collected metadata
-	dedupeStatementMetadata(statementMetadata)
-
 	return normalizer.trimNormalizedSQL(normalizedSQL), statementMetadata, nil
```

Comment:
This processing is very similar to [normalize.go](https://github.com/DataDog/go-sqllexer/pull/51/files#diff-4c348302fdfdc50fe2f22b11fd1214a0d23d90dcd2a79f75f08afa4ac15d2d2fR204-R219). Shall we avoid repetition?